scrapy_poet/repository.py

```diff
@@ -0,0 +1,55 @@
+"""The Injection Middleware needs a standard way to build dependencies for
+the Page Inputs used by the request callbacks. That's why we have created a
+repository of ``PageObjectInputProvider`` subclasses.
+
+You could implement different providers in order to acquire data from multiple
+external sources, for example, Splash or Auto Extract API.
+"""
+import typing
+
+from scrapy_poet import autoextract
+from scrapy_poet.providers import (
+    PageObjectInputProvider,
+    ResponseDataProvider,
+)
+# FIXME: refactor _providers / provides / register,  make a nicer API
+
+providers = {}
+
+
+def register(provider_class: typing.Type[PageObjectInputProvider],
+             provided_class: typing.Optional[typing.Type] = None):
+    """This method registers a Page Object Input Provider in the providers
+    registry. It could be replaced by the use of the ``provides`` decorator.
+
+    If ``provided_class`` is not specified, we get it from ``provider_class``.
```

Comment:
Maybe it makes sense to force the providers to have `provided_class`  always defined. And then change the decorator `provides` by `provider`. And the decorator also looks for `provided_class`. This way you could remove the lines `register(...)` below and replace them by a decorator in top of the provider classes.

Reply:
It's very tempting. But see that the main reason for calling `register` bellow is to avoid circular imports and weird usages like having to import classes or modules without using them just to make the providers register themselves through the use of a decorator. 

@kmike has suggested making it somehow mandatory like we having a place on our spider settings where we should list and register all required providers. This is actually following the opposite path of your suggestion. 

I think we should discuss pros and cons together.

Reply:
>  like having to import classes or modules without using them just to make the providers register themselves

I understand. Makes sense. But this does not affect the idea of having the property `provided_class` always defined, right?

> Having a place on our spider settings where we should list and register all required providers

Yes, I like something like that and is in sync with my idea exposed here: https://github.com/scrapinghub/scrapy-poet/pull/34#discussion_r466527309

Reply:
> I understand. Makes sense. But this does not affect the idea of having the property provided_class always defined, right?

You're right. I've just refactored this decorator and the register method as well. Could you check it, please?

Reply:
Checked. I left some comments.